pkg/virt-api/validating-webhook/validating-webhook.go

```diff
@@ -0,0 +1,456 @@
+	}
+	for idx, volume := range volumes {
+		// verify name is unique
+		otherIdx, ok := nameMap[volume.Name]
+		if !ok {
+			nameMap[volume.Name] = idx
+		} else {
+			errors = append(errors, fmt.Errorf("spec.volumes[%d] and spec.volumes[%d] must not have the same Name.", idx, otherIdx))
+		}
+
+		// Verify exactly one source is set
+		volumeSourceSetCount := 0
+		if volume.PersistentVolumeClaim != nil {
+			volumeSourceSetCount++
+		}
+		if volume.CloudInitNoCloud != nil {
+			volumeSourceSetCount++
+		}
+		if volume.RegistryDisk != nil {
+			volumeSourceSetCount++
+		}
+		if volume.Ephemeral != nil {
+			volumeSourceSetCount++
+		}
+		if volume.EmptyDisk != nil {
+			volumeSourceSetCount++
+		}
+
+		if volumeSourceSetCount != 1 {
+			errors = append(errors, fmt.Errorf("spec.volumes[%d] must have exactly one source type set", idx))
+		}
+
+		// Verify cloud init data is within size limits
+		if volume.CloudInitNoCloud != nil {
+			noCloud := volume.CloudInitNoCloud
+			userDataLen := 0
+
+			userDataSourceCount := 0
+			if noCloud.UserDataSecretRef != nil && noCloud.UserDataSecretRef.Name != "" {
+				userDataSourceCount++
+			}
+			if noCloud.UserDataBase64 != "" {
+				userDataSourceCount++
+				userData, err := base64.StdEncoding.DecodeString(noCloud.UserDataBase64)
+				if err != nil {
+					errors = append(errors, fmt.Errorf("spec.volumes[%d].cloudInitNoCloud.userDataBase64 is not a valid base64 value.", idx))
+				}
+				userDataLen = len(userData)
+			}
+			if noCloud.UserData != "" {
+				userDataSourceCount++
+				userDataLen = len(noCloud.UserData)
+			}
+
+			if userDataSourceCount != 1 {
+				errors = append(errors, fmt.Errorf("spec.volumes[%d].cloudInitNoCloud must have one exactly one userdata source set.", idx))
+			}
+
+			if userDataLen > cloudInitMaxLen {
+				errors = append(errors, fmt.Errorf("spec.volumes[%d].cloudInitNoCloud userdata exceeds %d byte limit", idx, cloudInitMaxLen))
+			}
+		}
+	}
+	return errors
+}
+
+func validateDevices(devices *v1.Devices) []error {
+	errors := []error{}
+	errors = append(errors, validateDisks(devices.Disks)...)
+	return errors
+}
+
+func validateDomainSpec(spec *v1.DomainSpec) []error {
+	errors := []error{}
+	errors = append(errors, validateDevices(&spec.Devices)...)
+	return errors
+}
+
+func validateVirtualMachineSpec(spec *v1.VirtualMachineSpec) []error {
+	errors := []error{}
+	volumeNameMap := make(map[string]int)
+
+	if len(spec.Domain.Devices.Disks) > arrayLenMax {
+		errors = append(errors, fmt.Errorf("spec.domain.devices.disks list exceeds the %d element limit in length", arrayLenMax))
+		// We won't process anything over the limit
+		return errors
+	} else if len(spec.Volumes) > arrayLenMax {
+		errors = append(errors, fmt.Errorf("spec.volumes list exceeds the %d element limit in length", arrayLenMax))
+		// We won't process anything over the limit
+		return errors
+	}
+
+	// Validate disks and VolumeNames match up correctly
+	for idx, disk := range spec.Domain.Devices.Disks {
```

Comment:
Here we still have the nested for-loop. I would prefer combining it with maps. Let's suppose we have 255 disk and validate 10 vms at the same time, we could theoretically end up with  `256*256*10 = 655360` comparisons (can happen easily deliberately or by accident if somes scripts are broken). We can do all necessary checks with a complexity of n by populating two maps. Worst: `3*256*10=7680` ( I think we need 2 full loops for populating the map and 1 for additional comparisions)  + two maps with 2*256 entries (not much memory).

Reply:
removing the nested loop works for me, fixed.